### PR TITLE
Week 8

### PR DESCRIPTION
**Pull request by alejandrodss**
On this PR:
* Adding SQL database and working with it.
* Add new scripts for setting the database
* For running the project:
  * npm run build
  * podman compose up -d
  * npm run sql-init-db
  * npm run start

### CODE REVIEW

File: src/order/order.repository.sql.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    for(const item of items) {
+      const orderItem = new OrderItem({
+        product: item.product.getEntity(),
+        orderId: order.id,
+        count: item.count
+      });
+      orderItems.push(orderItem);
+    }
+    order.items.load();
+    order.items.set(orderItems);
```
Comment (alejandrodss) [inline src/order/order.repository.sql.ts:62]: I'm setting the items on the order, but for some reason they are not being persisted on the DB.
Do you know what is happening there?

File: src/cart/cart.service.mongo.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
+import { BaseError } from "../exceptions/BaseError";
```
Comment (alejandrodss) [inline src/cart/cart.service.mongo.ts:1]: Please, ignore the files with mongo in their names. They are from the previous module. I just renamed them for keeping into the repository.

File: src/order/order.repository.sql.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    try {
+      const total = calculateTotalSql(order.items as CartItem[]);
+      const newOrder = new OrderEntity(
+        order.userId,
+        order.cartId,
+        total
+      );
+      const savedOrder = await this.orderRepository.insert(newOrder);
+      this._setItems(order.items as CartItem[], newOrder);
+      console.log("Order created: ", savedOrder);
```
Comment (carboleda) [inline src/order/order.repository.sql.ts]:
```suggestion
      const newOrder = new OrderEntity(order.userId, order.cartId, total);
      this._setItems(order.items as CartItem[], newOrder);
      console.log("newOrder", newOrder);
      await this.orderRepository.getEntityManager().persistAndFlush(newOrder);
      return newOrder;
```

File: src/order/order.repository.sql.ts (hunk 1 above)
Comment (carboleda) [inline src/order/order.repository.sql.ts:63]:
```suggestion
    order.items.set(orderItems);
```

File: src/order/order.repository.sql.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      const orderItem = new OrderItem({
+        product: item.product.getEntity(),
+        orderId: order.id,
+        count: item.count
+      });
+      orderItems.push(orderItem);
+    }
+    order.items.load();
+    order.items.set(orderItems);
+    this.orderRepository.getEntityManager().flush();
```
Comment (carboleda) [inline src/order/order.repository.sql.ts]:
```suggestion
    order.items.set(orderItems);
```

File: src/order/order.repository.sql.ts (hunk 1 above)
Comment (carboleda) [reply src/order/order.repository.sql.ts:62]: The suggestions I left in this file solved the issue of persisting the items

File: src/entities/orderItem.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { Entity, ManyToOne, Property, Ref, Reference } from "@mikro-orm/core";
+import { Order } from "./order";
+import { Product } from "./product";
+
+@Entity()
+export class OrderItem {
+  @ManyToOne(() => Order, {primary: true, ref: true})
+  order!: Ref<Order>;
```
Comment (carboleda) [inline src/entities/orderItem.ts:8]: M: It looks like you wrongly used the order as the primary key here, and this is failing when you try to create an order with more than 2 items. You should probably create an id for this entity or use a compound key (`orderId`, `productId`)

File: src/entities/orderItem.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  description: string;
+
+  @Property({ columnType: 'float'})
+  price: number;
+
+  @Property()
+  count: number;
+
+  constructor(dto: { product: Product, orderId: string, count: number}) {
+    this.id = dto.product.id;
```
Comment (carboleda) [inline src/entities/orderItem.ts]: S: If the `OrderItem.id` is actually the productId it would be better to name it as `productId`

**Review by carboleda — CHANGES_REQUESTED** [review]
I left you a couple of comments that need to be addressed

File: src/order/order.repository.sql.ts (hunk 1 above)
Comment (alejandrodss) [reply src/order/order.repository.sql.ts:62]: I just tested and it's working.
Thank you!

File: src/entities/orderItem.ts (hunk 5 above)
Comment (alejandrodss) [reply src/entities/orderItem.ts:8]: Fixed!

**Review by carboleda — APPROVED** [review]
Good job! 🚀

**Pull request merged**